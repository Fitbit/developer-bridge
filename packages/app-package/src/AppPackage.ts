import aprMap from 'apr-map';
import jszip = require('jszip');
import lodash = require('lodash');

/**
 * The source maps are not currently validated, so it would be
 * misleading to claim that it is a valid source map.
 */
type RawSourceMap = unknown;

interface PackageComponents {
  device: {
    [family: string]: {
      artifact: Buffer,
      platform?: string[],
    },
  };
  companion?: Buffer;
}

interface DeviceSourceMaps {
  [family: string]: ComponentSourceMaps;
}

export interface SourceMaps {
  device: DeviceSourceMaps;
  companion?: ComponentSourceMaps;
  settings?: ComponentSourceMaps;
}

export interface ComponentSourceMaps {
  [filename: string]: RawSourceMap;
}

interface BundledComponentSourceMaps {
  [filename: string]: string;
}

interface ManifestSourceMaps {
  device: {
    [family: string]: BundledComponentSourceMaps;
  };
  companion?: BundledComponentSourceMaps;
  settings?: BundledComponentSourceMaps;
}

interface ManifestCommon {
  manifestVersion: number;
  buildId: string;
  appId: string;
  requestedPermissions: string[];
  components: {
    companion?: {};
  };
}

interface ManifestV5 extends ManifestCommon {
  manifestVersion: 5;
  components: {
    watch: string;
    companion?: string;
  };
  platform: string[];
}

interface ManifestV6 extends ManifestCommon {
  manifestVersion: 6;
  sdkVersion: SDKVersion;
  components: {
    watch: {
      [family: string]: {
        platform?: string[];
        filename: string;
      };
    };
    companion?: {
      filename: string;
    };
  };
  sourceMaps?: ManifestSourceMaps;
}

interface SDKVersion {
  deviceApi: string;
  companionApi?: string;
}

function getFile(zip: jszip, path: string) {
  const file = zip.file(path);
  if (!file || file.dir) throw new Error(`${path} not present in zip file`);
  return file;
}

async function getBufferFromZip(zip: jszip, path: string) {
  return getFile(zip, path).async('nodebuffer');
}

async function getTextFromZip(zip: jszip, path: string) {
  return getFile(zip, path).async('text');
}

const extractComponentSourceMaps =
  (sourceMapPaths: BundledComponentSourceMaps, zip: jszip): Promise<ComponentSourceMaps> =>
    aprMap(sourceMapPaths, path => getTextFromZip(zip, path).then(JSON.parse));

async function extractSourceMaps(zip: jszip, sourceMapManifest?: ManifestSourceMaps) {
  if (!sourceMapManifest) return undefined;

  const extractComponent = (component?: BundledComponentSourceMaps) =>
    component && extractComponentSourceMaps(component, zip);

  return {
    device: await aprMap(
      sourceMapManifest.device,
      component => extractComponentSourceMaps(component, zip),
    ) as DeviceSourceMaps,
    companion: await extractComponent(sourceMapManifest.companion),
    settings: await extractComponent(sourceMapManifest.settings),
  };
}

abstract class ManifestParserBase {
  protected manifest!: ManifestCommon;

  abstract getSourceMapExtractor(): (zip: jszip) => Promise<SourceMaps | undefined>;

  pullMetadata() {
    return {
      buildId: this.manifest.buildId,
      uuid: this.manifest.appId,
      requestedPermissions: this.manifest.requestedPermissions,
    };
  }

  getSDKVersions() {
    const versions: SDKVersion = {
      deviceApi: '1.0.0',
    };
    if (this.manifest.components.companion) versions.companionApi = '1.0.0';
    return versions;
  }
}

class ManifestParserV5 extends ManifestParserBase {
  constructor(protected manifest: ManifestV5) {
    super();
  }

  getDeviceComponents() {
    if (typeof this.manifest.components !== 'object' || !this.manifest.components.watch) {
      throw new Error('No components listed in manifest.json');
    }

    if (!Array.isArray(this.manifest.platform)) {
      throw new Error('Missing platform descriptors');
    }

    return this.manifest.platform.map<[string, { filename: string, platform?: string[] }]>(
      (platformDescriptor) => {
        const [, family, platform] = /^([^:]+):?(.+)?$/.exec(platformDescriptor)!;

        return [
          family.toLowerCase(),
          {
            platform: platform ? [platform] : undefined,
            filename: this.manifest.components.watch,
          },
        ];
      },
    );
  }

  getCompanionFilename() {
    return this.manifest.components.companion;
  }

  getSourceMapExtractor() {
    return () => Promise.resolve(undefined);
  }
}

class ManifestParserV6 extends ManifestParserBase {
  constructor(protected manifest: ManifestV6) {
    super();
  }

  getDeviceComponents() {
    if (typeof this.manifest.components !== 'object') {
      throw new Error('No components listed in manifest.json');
    }

    if (!this.manifest.components.watch) return [];

    return Object.entries(this.manifest.components.watch);
  }

  getCompanionFilename() {
    return lodash.get(this.manifest, 'components.companion.filename');
  }

  getSDKVersions() {
    return {
      ...super.getSDKVersions(),
      ...this.manifest.sdkVersion,
    };
  }

  getSourceMapExtractor() {
    return (zip: jszip) => extractSourceMaps(zip, this.manifest.sourceMaps);
  }
}

function getManifestParser(manifest: ManifestCommon) {
  switch (manifest.manifestVersion) {
    case 5:
      return new ManifestParserV5(manifest as ManifestV5);
    case 6:
      return new ManifestParserV6(manifest as ManifestV6);
    default:
      throw new Error(`Unsupported manifest version ${manifest.manifestVersion}`);
  }
}

export interface AppPackage {
  buildId: string;
  components: PackageComponents;
  sourceMaps?: SourceMaps;
  requestedPermissions: string[];
  uuid: string;
  sdkVersion: SDKVersion;
}

export async function fromJSZip(fbaZip: jszip): Promise<AppPackage> {
  const textFile = (path: string) => getTextFromZip(fbaZip, path);
  const bufferFile = (path: string) => getBufferFromZip(fbaZip, path);

  const manifestJSON = JSON.parse(await textFile('manifest.json'));
  const parser = getManifestParser(manifestJSON);

  const device = await Promise.all(
    parser.getDeviceComponents().map(
      ([family, { platform, filename }]) => bufferFile(filename)
        .then(artifact => [family, { platform, artifact }]),
    ),
  ).then(lodash.fromPairs);

  const companionFilename = parser.getCompanionFilename();
  const companion = companionFilename ? await bufferFile(companionFilename) : undefined;

  const sourceMaps = await parser.getSourceMapExtractor()(fbaZip);

  return {
    ...parser.pullMetadata(),
    sourceMaps,
    components: {
      device,
      companion,
    },
    sdkVersion: parser.getSDKVersions(),
  };
}
