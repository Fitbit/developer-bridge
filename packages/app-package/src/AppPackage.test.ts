import jszip = require('jszip');

import { fromJSZip } from './AppPackage';

const manifestCommon = {
  appId: 'eff9d309-eadd-41d7-9af3-696eafc3fa31',
  buildId: '0x0123456789abcdef',
};

describe('fromJSZip', () => {
  const manifestV5 = {
    ...manifestCommon,
    manifestVersion: 5,
    platform: [],
  };

  const manifestV6 = {
    ...manifestCommon,
    manifestVersion: 6,
  };

  type FilesObject = { [path: string]: object | string };

  function buildPackage(files: FilesObject = {}) {
    const zip = new jszip();
    for (const [path, content] of Object.entries(files)) {
      zip.file(
        path,
        typeof content === 'object' ? JSON.stringify(content) : content,
      );
    }
    return zip;
  }

  function itRejects(name: string, files: FilesObject = {}) {
    it(`rejects ${name}`, async () =>
      expect(fromJSZip(await buildPackage(files))).rejects.toMatchSnapshot());
  }

  function itAccepts(name: string, files: FilesObject) {
    it(`accepts ${name}`, async () =>
      expect(fromJSZip(await buildPackage(files))).resolves.toMatchSnapshot());
  }

  itRejects('a zip file without a manifest');

  itRejects('a manifest that is not JSON', {
    'manifest.json': 'Not JSON',
  });

  itRejects('a manifest with unsupported version', {
    'manifest.json': { manifestVersion: 100 },
  });

  [5, 6].forEach((version) => {
    itRejects(`a v${version} manifest with mistyped components`, {
      'manifest.json': {
        manifestVersion: version,
        components: 'components!',
      },
    });
  });

  itRejects('a v5 manifest with no device component', {
    'manifest.json': {
      ...manifestV5,
      components: { companion: 'companion.zip' },
    },
    'companion.zip': '',
  });

  itRejects(
    'a v5 manifest which references a nonexistent device component file',
    {
      'manifest.json': {
        ...manifestV5,
        platform: ['HIGGS'],
        components: { watch: 'doesnotexist.zip' },
      },
    },
  );

  itRejects(
    'a v6 manifest which references a nonexistent device component file',
    {
      'manifest.json': {
        ...manifestV6,
        components: {
          watch: {
            higgs: { filename: 'doesnotexist.zip' },
          },
        },
      },
    },
  );

  itRejects(
    'a v5 manifest which references a nonexistent companion component file',
    {
      'manifest.json': {
        ...manifestV5,
        components: {
          watch: 'device.zip',
          companion: 'companion.zip',
        },
      },
      'device.zip': '',
    },
  );

  itRejects(
    'a v6 manifest which references a nonexistent companion component file',
    {
      'manifest.json': {
        ...manifestV6,
        components: {
          watch: {
            higgs: { filename: 'higgs.zip' },
          },
          companion: { filename: 'companion.zip' },
        },
      },
      'higgs.zip': '',
    },
  );

  itRejects('a v5 manifest with no platform descriptor', {
    'manifest.json': {
      ...manifestV5,
      platform: undefined,
      components: { watch: 'device.zip' },
    },
    'device.zip': '',
  });

  itAccepts('a v5 manifest with an unversioned platform', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v5 manifest with a versioned platform', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS:32.1.16+'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v5 manifest with multiple platforms', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS:32.1.16+', 'MESON:1.2.3+'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v6 manifest with no platform spec for the device', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: {
          higgs: { filename: 'higgs.zip' },
        },
      },
    },
    'higgs.zip': 'higgs.zip contents',
  });

  itAccepts('a v6 manifest with multiple device components', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: {
          higgs: {
            platform: ['23.1.5', '32.4.16+'],
            filename: 'higgs.zip',
          },
          meson: {
            platform: ['32.4.17+'],
            filename: 'meson.zip',
          },
        },
      },
    },
    'higgs.zip': 'higgs',
    'meson.zip': 'meson',
  });

  itAccepts('a v5 manifest with a companion', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS'],
      components: {
        watch: 'device.zip',
        companion: 'companion.zip',
      },
    },
    'device.zip': 'device',
    'companion.zip': 'companion',
  });

  itAccepts('a v6 manifest with a companion', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: { higgs: { filename: 'higgs.zip' } },
        companion: { filename: 'companion.zip' },
      },
    },
    'higgs.zip': 'higgs',
    'companion.zip': 'companion',
  });

  itAccepts('a v6 manifest with no device component', {
    'manifest.json': {
      ...manifestV6,
      components: {
        companion: { filename: 'companion.zip' },
      },
    },
    'companion.zip': '',
  });

  itAccepts('a v6 manifest with sourcemaps but no device component', {
    'manifest.json': {
      ...manifestV6,
      sourceMaps: {
        companion: {},
      },
      components: {
        companion: { filename: 'companion.zip' },
      },
    },
    'companion.zip': '',
  });
});
