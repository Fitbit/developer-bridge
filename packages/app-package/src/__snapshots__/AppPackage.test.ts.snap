// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromJSZip accepts a v5 manifest with a companion 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [
        99,
        111,
        109,
        112,
        97,
        110,
        105,
        111,
        110,
      ],
      "type": "Buffer",
    },
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v5 manifest with a versioned platform 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.1.16+",
        ],
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v5 manifest with an unversioned platform 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v5 manifest with multiple platforms 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.1.16+",
        ],
      },
      "meson": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "1.2.3+",
        ],
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v6 manifest with a companion 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [
        99,
        111,
        109,
        112,
        97,
        110,
        105,
        111,
        110,
      ],
      "type": "Buffer",
    },
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v6 manifest with multiple device components 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "23.1.5",
          "32.4.16+",
        ],
      },
      "meson": Object {
        "artifact": Object {
          "data": Array [
            109,
            101,
            115,
            111,
            110,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.4.17+",
        ],
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v6 manifest with no device component 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "device": Object {},
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip accepts a v6 manifest with no platform spec for the device 1`] = `
Object {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromJSZip rejects a manifest that is not JSON 1`] = `[SyntaxError: Unexpected token N in JSON at position 0]`;

exports[`fromJSZip rejects a manifest with unsupported version 1`] = `[Error: Unsupported manifest version 100]`;

exports[`fromJSZip rejects a v5 manifest which references a nonexistent companion component file 1`] = `[Error: companion.zip not present in zip file]`;

exports[`fromJSZip rejects a v5 manifest which references a nonexistent device component file 1`] = `[Error: doesnotexist.zip not present in zip file]`;

exports[`fromJSZip rejects a v5 manifest with mistyped components 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromJSZip rejects a v5 manifest with no device component 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromJSZip rejects a v5 manifest with no platform descriptor 1`] = `[Error: Missing platform descriptors]`;

exports[`fromJSZip rejects a v6 manifest which references a nonexistent companion component file 1`] = `[Error: companion.zip not present in zip file]`;

exports[`fromJSZip rejects a v6 manifest which references a nonexistent device component file 1`] = `[Error: doesnotexist.zip not present in zip file]`;

exports[`fromJSZip rejects a v6 manifest with mistyped components 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromJSZip rejects a zip file without a manifest 1`] = `[Error: manifest.json not present in zip file]`;
